enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,246 @@
+resolves the name. Otherwise, CoreDNS proxies DNS messages to an upstream resolver identified by `/etc/resolv.conf` on
+the corresponding node. Although this provides a consistent and reliable approach for name resolution, it restricts how
+cluster operators manage DNS name queries.
+
+### Goals
+
+1. A well defined API for DNS forwarding configuration that is implementation agnostic.
+2. A minimal API surface that can be expanded to support future DNS forwarding use cases.
+
+### Non-Goals
+
+1. Support every possible DNS forwarding use case.
+2. Configure or manage external DNS providers.
+
+## Proposal
+
+The proposed `Upstream` type defines a schema for associating one or more `Nameservers` to a DNS subdomain identified by
+`Domain`. `Nameservers` are responsible for resolving name queries for `Domain`:
+```go
+type Upstream struct {
+	Domain string `json:"domain"`
+	Nameservers []string `json:"nameservers"`
+}
+```
+
+Each instance of CoreDNS performs health checking of `Nameservers`. If `Nameservers` consists of more than one
+nameserver, forwarding will be randomized on first use. When a healthy nameserver returns an error during the exchange,
+the next server in the list is tried. Each server is represented by an IP address or IP address and port if the server
+listens on a UDP port other than 53.
+
+A list of `Upstream` is added to `dnses.opertor.openshift.io`, allowing users the ability to configure multiple
+upstream resolvers to subdomain associations:
+
+```go
+// DNSSpec is the specification of the desired behavior of the DNS.
+type DNSSpec struct {
+	// +optional
+	Upstreams []Upstream `json:"upstreams,omitempty"`
+}
+```
+
+### User Stories
+
+#### Story 1
+
+As a customer with OpenShift running in AWS and connected to my data center by VPN, I want OpenShift DNS to resolve
+name queries for our other internal devices using the DNS servers in our data center.
+
+### Implementation Details/Notes/Constraints
+
+The maximum number of `Nameservers` is 15.
+
+The cluster-dns-operator will generate the `forward` portion of `configmap/dns-default` based on `Upstream` of
+`dnses.operator.openshift.io` instead of `forward` being statically defined. To achieve this:
+
+1. The [default](https://github.com/openshift/cluster-dns-operator/blob/master/assets/dns/configmap.yaml) ConfigMap
+asset and associated code from the [manifests pkg](https://github.com/openshift/cluster-dns-operator/blob/master//pkg/manifests/manifests.go#L70:6)
+should be removed.
+
+2. The [desiredDNSConfigMap](https://github.com/openshift/cluster-dns-operator/blob/master/pkg/operator/controller/controller_dns_configmap.go)
+function must be modified to create a ConfigMap type with a `forward` configuration based on `Upstream`.
+If `Upstream` is not present or an invalid `Upstream` is provided, the ConfigMap will contain
+`forward . /etc/resolv.conf`. Otherwise, desiredDNSConfigMap will use the provided `Upstream` to construct the
+forwarding configuration.
```

Comment:
Does this mean that using `upstreams` is mutually exclusive with forwarding to `/etc/resolv.conf`?

That is, if `upstreams` are specified, `upstreams` represents the totality of the forwarding chain and `/etc/resolv.conf` will _not be honored_?

Reply:
> That is, if upstreams are specified, upstreams represents the totality of the forwarding chain and /etc/resolv.conf will not be honored?

@ironcladlou yes. Instead, should we have the following prepended to any user-defined upstreams?

```
forward . /etc/resolv.conf {
        policy sequential
}
```

Reply:
> > That is, if upstreams are specified, [...] /etc/resolv.conf will not be honored?
> 
> @ironcladlou yes.

That is surprising to me, and it could be nasty surprise for administrators.

> Instead, should we have the following prepended to any user-defined upstreams?
> 
> ```
> forward . /etc/resolv.conf {
>         policy sequential
> }
> ```

I vote for this.

Reply:
Will the delegation order of resolv.conf and user-defined upstreams become significant in this case? Some options:

1. Choose a static ordering
2. Add an ordering configuration

Reply:
Per https://coredns.io/manual/toc/#configuration,

> If there are multiple Servers configured that listen on the queried port, it will check which one has the most specific zone for this query (longest suffix match). E.g. if there are two Servers, one for `example.org` and one for `a.example.org`, and the query is for `www.a.example.org`, it will be routed to the latter.

This seems like reasonable semantics.

Reply:
> Will the delegation order of resolv.conf and user-defined upstreams become significant in this case?

I've added the `UpstreamPolicy` type to the proposed API, allowing users to specify the upstream selection policy with a default of `Sequential`. Let me know if you prefer `Random` as the default to provide consistency with upstream. __Note:__ The `resolv.conf` resolver of last resort will use policy `Sequential` independent of any `Upstream` policy.